File: src/main/scala/mod/iceandshadow3/lib/compat/block/type/AProperty.scala
package mod.iceandshadow3.lib.compat.block.`type`

import java.util.Optional

import mod.iceandshadow3.lib.block.BlockVar
import mod.iceandshadow3.lib.util.OrdSet
import net.minecraft.state.IProperty

import scala.jdk.CollectionConverters._

class AProperty(logic: BlockVar[_]) extends IProperty[Integer] {
	override val getAllowedValues = new OrdSet(logic.size).asJava

	override def getName = logic.name
	override def getValueClass = classOf[Integer]
	override def parseValue(value: String) = {
		val int = logic.parseHalf(value)
		if(int == -1) Optional.empty() else Optional.of(int)
	}
	override def getName(value: Integer) = logic.toString(value)

	override def toString = logic.name+" (AProperty["+logic.size+"])"
}

File: src/main/scala/mod/iceandshadow3/lib/compat/block/impl/AProperty.java
package mod.iceandshadow3.lib.compat.block.impl;

import mod.iceandshadow3.lib.util.collect.OrdSet;
import net.minecraft.state.IProperty;

import javax.annotation.Nonnull;
import java.util.Collection;
import java.util.Optional;

public class AProperty implements IProperty<Integer> {
	private final BVarBlockNew logic;
	private final Collection<Integer> values;

	AProperty(BVarBlockNew<?> bbv) {
		logic = bbv;
		values = new OrdSet(logic.size()).toJava();
	}

	@Override
	@Nonnull
	public String getName() {
		return logic.name();
	}

	@Override
	@Nonnull
	public Collection<Integer> getAllowedValues() {
		return values;
	}

	@Override
	@Nonnull
	public Class<Integer> getValueClass() {
		return Integer.class;
	}

	@Override
	@Nonnull
	public Optional<Integer> parseValue(@Nonnull String value) {
		final int idx = logic.fromString(value);
		return idx != -1 ? Optional.of(idx) : Optional.empty();
	}

	@Override
	@Nonnull
	public String getName(@Nonnull Integer value) {
		return logic.toString(value);
	}
}
